### ISSUE TITLE
[META][PERF] Performance issues, docs, benchmarking and improvements

### ISSUE THREAD

**Issue by man-shu** [nilearn/nilearn#5966]
### Issues
Several performance-related issues have been reported in Nilearn; these can be filtered using the "Performance" label here: https://github.com/nilearn/nilearn/issues?q=is:issue%20state:open%20label:Performance. But here is a list to keep track of things:
- #719 
- #1362 
- #1754
- #1883
- #2209
- #3177 
- #3592 
- #4811
- #5588
- https://neurostars.org/t/problem-with-smoothing-in-first-level-analysis/29553

### Documentation
There have also been efforts to understand/explain these issues. These are mostly the issues/PRs where we have used profilers to look at memory usage under-the-hood. So far [memory_profiler](https://github.com/pythonprofilers/memory_profiler) has been most useful because it can be used within ipython/within scripts/via command line and also allows tracking parallel processes:
- #3398
- #3399
- #3401 
- #4811
- We have a high-level example comparing different strategies to do masking in parallel: #5113 and #5185
- We noticed from the above example that memory usage in parallel depended on how the image is initially loaded. So we started to understand and explain the differences between nilearn vs. nibabel loading functions and consequently proxy vs. array images in #5128. At the time of writing this issue, in this PR, we are stuck at why the actual memory usage multiplier is higher than the number of parallel processes.

### Performance benchmarking
It also makes sense to track both computation time and memory usage over time, flag any regressions and have a way to reliably test proposed performance improvements. For now we are following sklearn (https://scikit-learn.org/dev/developers/contributing.html#monitoring-performance) in using [asv](https://asv.readthedocs.io/en/latest/) for this purpose
- #4573 
- Initial placeholder benchmarks were implemented in #5280 and the code lives in [asv_benchmarks](https://github.com/nilearn/nilearn/tree/main/asv_benchmarks) subdirectory.
- Reports from these benchmarks are deployed here: https://nilearn.github.io/benchmarks/
- Now we want to reorganize [asv_benchmarks](https://github.com/nilearn/nilearn/tree/main/asv_benchmarks) submodule-wise in #5590. 
- It could also be worthwhile to also have a distinction between benchmarks that are light on resources so that they can be run on regular intervals to flag any regressions vs. the ones that are resource-intensive and should/could only be run on specific PRs or locally.

### Improvements
This list is small but could grow larger as we understand the issues better. For now, one possibility is just to allow GPU usage via nilearn. Here also we could follow what sklearn is doing by introducing Array API support (see [sklearn 1.8 release highlights](https://scikit-learn.org/stable/auto_examples/release_highlights/plot_release_highlights_1_8_0.html#array-api-support-enables-gpu-computations) and https://scikit-learn.org/stable/modules/array_api.html)
- #3401
- #5753
- #5958

**Comment by man-shu** [comment]
This is just an initial draft for the state of affairs for Nilearn's performance from my POV. Please feel free to edit/comment as you prefer.

**Comment by bthirion** [comment]
I think it would be great to identify one aspect to prioritize.

**Comment by Remi-Gau** [comment]
## Improvements when running benchmarks locally

- [x] https://github.com/nilearn/nilearn/pull/5590#discussion_r2824280273
- [ ] make it easier to tweak the benchmarks to run them locally either with a lot of voxels (high res) or a lot volumes (long acquisitions)
- [ ] ensure that benchmark website is updated

**Comment by Remi-Gau** [comment]
All benchmarks are run on every PR even when the PR did not touch some of the code the benchmarks are concerned with: we should try to reuse the logic of only running the necessary tests (via the architecture enforced by importlinter and build_tools/github/restrict_tests_to_run.py) and apply it to the benchmarks

**Comment by bthirion** [comment]
You're obviously right.
